### DISCUSSION TITLE
How to customize user session data

### DISCUSSION THREAD

**Discussion by dsebastien** [supabase/supabase#14708] [Questions] [answered]
In my application, I don't only rely on the `auth.users` table, but also on an internal users table: `public.users`. When I call `supabaseClient.auth.getSession().user` I can access the user data coming from auth.users. I suppose that information is extracted from a JWT token created after authentication.

I need information from my internal users table in various situations (e.g., to customize the UI, to create links between entities and the current user, knowing that this relies on the internal user id).

Is there a straightforward way to:
- Customize the JWT tokens generated by supabase to include custom claims in the generated JWT tokens?
- Include additional data in the user session?

If not, how would you recommend approaching this use case?

**Comment by dsebastien** [comment, marked as answer]
I've found a solution that seems to work.

I have a PGPLSQL function that is called by a trigger whenever a row is added to the auth.users table. That function is the one that creates a row in my public.users table. Now, after generating an internal id for the new user, it modifies the `raw_user_metadata` column to include that information:

```
drop trigger if exists on_auth_user_created on auth.users;
create trigger on_auth_user_created
  after insert
  on auth.users
  for each row
  execute procedure public.handle_new_user();

create or replace function public.handle_new_user()
returns trigger
security definer
language plpgsql
as
$$
declare
...
begin
...
select gen_random_uuid()
into internal_user_id;

insert into public.users (id, user_id_external, username, email)
values (internal_user_id, new.id, user_name, new.email);

...

-- CUSTOMIZE THE APP METADATA
-- Store the internal user id in the app metadata
-- Thanks to this, we can retrieve it through the JWT and through the Supabase session
update auth.users set raw_user_meta_data = raw_user_meta_data || json_build_object('user_id_internal', internal_user_id)::jsonb where id = new.id;

return new;
end;
$$;
```

Since the `raw_user_meta_data` is included in the JWT tokens and accessible via the supabase client, it means I can access that information wherever I can retrieve a supabase session, which is exactly what I needed.

The only drawback of this approach is that the JWT token generated upon signup does NOT contain my updated information (it is generated too soon it seems). But it works after a logout/login (since the token is then refreshed). One workaround is to force a refresh of the JWT token after signup/login. I did that like this on the page users end up after login:

```
export const getServerSideProps: GetServerSideProps<Partial<CustomPageProps> & DashboardPageProps> = async (ctx) => {
  ...

  // Refreshing the session to make sure it includes what we need
  // WARNING: This is mandatory because of the customizations we make to app_meta_data and user_meta_data
  // which are not immediately reflected in the JWT token generated after user sign up
  // We normally only need this once after signup, but we accept the tradeoff
  const supabaseClient = createServerSupabaseClient<Database>(ctx);
  await supabaseClient.auth.refreshSession();
  ...
```

**Reply by dsebastien** [reply]
Unfortunately, starting the app today, I get the following error:

```
AuthApiError: Invalid Refresh Token: Already Used
    at /home/sebastien/wks/knowii/node_modules/@supabase/gotrue-js/dist/main/lib/fetch.js:41:20
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  __isAuthError: true,
  status: 400

```

So I'm not there yet :p

**Reply by Hazemkak** [reply]
where you able to get there @dsebastien
